### PR TITLE
[Serve] use RequestProtocol to determine the protocol

### LINKED ISSUE #38193: [Serve] Productionize gRPC proxy work
### Description

We have explored all options and finalized interface for using the new gRPC proxy. Let's clean up the code and make sure all parts are tested. 

### Use case

_No response_

### PR DESCRIPTION
**Pull request by GeneDer**
## Why are these changes needed?

- Added `RequestProtocol` enum
- Removed `_is_for_http_requests` boolean flag and replaced with `_request_protocol` on the serve handle and `RequestMetadata`
- Renamed `proxy_name` on the proxy to `protocol`
- Fixed a small bug metrics not logged properly when chaining handles (by passing metrics to the handle chain)

## Related issue number

7th PR to close https://github.com/ray-project/ray/issues/38193

## Checks

- [ ] I've signed off every commit(by using the -s flag, i.e., `git commit -s`) in this PR.
- [ ] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
    - [ ] I've added any new APIs to the API Reference. For example, if I added a 
           method in Tune, I've added it in `doc/source/tune/api/` under the 
           corresponding `.rst` file.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

File: python/ray/serve/handle.py [hunk 1]
```diff
@@ -63,13 +63,15 @@ class HandleOptions:
...
     _router_cls: str = ""
+    request_protocol: RequestProtocol = RequestProtocol.UNDEFINED
 
     def copy_and_update(
         self,
         method_name: Union[str, DEFAULT] = DEFAULT.VALUE,
         multiplexed_model_id: Union[str, DEFAULT] = DEFAULT.VALUE,
         stream: Union[bool, DEFAULT] = DEFAULT.VALUE,
         _router_cls: Union[str, DEFAULT] = DEFAULT.VALUE,
+        request_protocol: Union[RequestProtocol, DEFAULT] = DEFAULT.VALUE,
```
Comment (edoakes) [inline python/ray/serve/handle.py:74]: let's make this a private/undocumented option please! (prefix with `_`)

Comment (edoakes) [reply python/ray/serve/handle.py:74]: actually even better would be to not put this in `.options()` and instead have a separate private method for setting it (`_set_request_protocol`). that way users aren't confused by it.

File: python/ray/serve/_private/router.py [hunk 2]
```diff
@@ -61,9 +65,6 @@ class RequestMetadata:
     endpoint: str
     call_method: str = "__call__"
 
-    # This flag is set if the request is made from the HTTP proxy to a replica.
-    is_http_request: bool = False
```
Comment (sihanwang41) [inline python/ray/serve/_private/router.py:65]: oh nice, make it more general. :)

**Review by sihanwang41 — APPROVED** [review]

File: python/ray/serve/handle.py [hunk 3]
```diff
@@ -193,12 +196,14 @@ def _options(
...
 
-        return self.__class__(
+        new_handle = self.__class__(
             self.deployment_name,
             handle_options=new_handle_options,
             _router=None if _router_cls != DEFAULT.VALUE else self._router,
-            _is_for_http_requests=self._is_for_http_requests,
+            _request_counter=self.request_counter,
         )
+        new_handle._set_request_protocol(self._request_protocol)
```
Comment (edoakes) [inline python/ray/serve/handle.py:205]: I would suggest still putting the protocol in the `HandleOptions` class (it's not publicly exposed). Then you don't need to have a separate call here.

Comment (GeneDer) [reply python/ray/serve/handle.py:205]: oh makes sense XD, let me fix it real quick

**Review by edoakes — APPROVED** [review]

**Pull request merged**